File: CPE/Exam/130326/pF_UVA-10308_Roads in the North.cpp
/*
    樹形 DP ，求直徑
*/
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    string line;
    while (getline(cin, line)) {
        int u, v, w;
        unordered_map<int, vector<pair<int, int>>> g;
        while (true) {
            if (line.empty()) break;
            stringstream ss(line);
            ss >> u >> v >> w;
            g[u].emplace_back(v, w);
            g[v].emplace_back(u, w);
            getline(cin, line);
        }

        int ans = 0;
        function<int(int, int)> dfs = [&](int u, int fa) -> int {
            int res1 = 0, res2 = 0;
            for (auto &[v, w] : g[u]) {
                if (v == fa) continue;
                int cur = w + dfs(v, u);
                if (cur > res1) {
                    res2 = res1;
                    res1 = cur;
                }
                else if (cur > res2) {
                    res2 = cur;
                }
            }
            ans = max(ans, res1 + res2);
            return res1;
        };
        dfs(1, -1);
        cout << ans << endl;
    }
    return 0;
}

File: CPE/Exam/130326/pF_UVA-10308_Roads in the North.py
"""
    樹形 DP ，求直徑
"""
from collections import defaultdict

run = True
while run:
    g = defaultdict(list)
    while True:
        try:
            line = input()
            if line == "":
                break
            u, v, w = map(int, line.split())
            g[u].append((v, w))
            g[v].append((u, w))
        except EOFError:
            run = False
            break
    
    ans = 0
    def dfs(u, fa):
        global ans
        res1, res2 = 0, 0
        for v, w in g[u]:
            if v == fa:
                continue
            cur = w + dfs(v, u)
            if cur > res1:
                res2 = res1
                res1 = cur
            elif cur > res2:
                res2 = cur
        ans = max(ans, res1 + res2)
        return res1
    
    dfs(1, -1)
    print(ans)